File: backend/app/routers/master.py
from fastapi import APIRouter, Depends, Query
from fastapi.responses import JSONResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies import AuthUser, ROLE_MASTER_ACCESS
from app.models import (
    LkAllowanceRates,
    LkBankMaster,
    LkGradeDerivation,
    LkGradeStep,
    LkLocationMaster,
    LkOrgMaster,
    LkPitBrackets,
)
from app.schemas.master import (
    AllowanceRateCreate,
    AllowanceRateOut,
    AllowanceRateUpdate,
    BankCreate,
    BankOut,
    BankUpdate,
    GradeDerivationCreate,
    GradeDerivationOut,
    GradeDerivationUpdate,
    GradeStepOut,
    GradeStepUpdate,
    LocationCreate,
    LocationOut,
    LocationUpdate,
    OrgCreate,
    OrgOut,
    OrgUpdate,
    PITBracketOut,
    PITBracketUpdate,
)
from app.services import master_service
from app.services.dashboard_service import cached_json
from app.services.master_service import (
    _allowance_rate_type,
    _bank_is_active,
    _location_is_active,
    _org_is_active,
    _pit_upper_display,
)

router = APIRouter(prefix="/master", tags=["master"])

_MASTER_EMPLOYEE_COUNTS_KEY = "master:employee-counts"


def _ok(data: dict | list) -> dict:
    return {"success": True, "data": data, "pagination": None, "error": None}


def _grade_step_out(row: LkGradeStep) -> dict:
    # Monthly basic = index points × LAK per point (matches payroll_service._calculate_one).
    basic = float(row.grade_step_index) * float(row.salary_index_rate)
    return GradeStepOut(
        grade=row.grade,
        step=row.step,
        basic_salary=basic,
    ).model_dump(mode="json")


def _allowance_out(row: LkAllowanceRates) -> dict:
    return AllowanceRateOut(
        allowance_name=row.allowance_name,
        rate_type=_allowance_rate_type(row.eligibility),
        rate_value=float(row.amount_or_rate),
        effective_date=row.effective_from,
    ).model_dump(mode="json")


def _grade_deriv_out(row: LkGradeDerivation) -> dict:
    return GradeDerivationOut(
        education_level=row.education_level,
        min_exp_years=row.exp_min_years,
        derived_grade=row.derived_grade,
        derived_step=row.derived_step,
    ).model_dump(mode="json")


def _org_out(row: LkOrgMaster) -> dict:
    return OrgOut(
        ministry_name=row.ministry_name,
        dept_key=row.department_key,
        dept_display_name=row.department_name,
        is_active=_org_is_active(row),
    ).model_dump(mode="json")


def _location_out(row: LkLocationMaster) -> dict:
    return LocationOut(
        province=row.province,
        region=row.district,
        is_remote_area=row.is_remote,
        is_hazardous_area=row.is_hazardous,
        is_active=_location_is_active(row),
    ).model_dump(mode="json")


def _bank_out(row: LkBankMaster) -> dict:
    swift = row.swift_code
    if swift == "000000000000":
        swift = None
    return BankOut(
        bank_name=row.bank_name,
        bank_code=row.branch_code,
        swift_code=swift,
        is_active=_bank_is_active(row),
    ).model_dump(mode="json")


def _pit_out(row: LkPitBrackets) -> dict:
    return PITBracketOut(
        bracket_no=row.bracket_no,
        lower_bound=float(row.income_from_lak),
        upper_bound=_pit_upper_display(row.income_to_lak),
        rate_pct=float(row.rate_pct),
        deduction_amount=float(row.cumulative_tax_lak),
    ).model_dump(mode="json")


@router.get("/grade-step")
async def get_grade_step(_user: AuthUser, db: AsyncSession = Depends(get_db)) -> dict:
    rows = await master_service.get_all_grade_steps(db)
    return _ok([_grade_step_out(r) for r in rows])


@router.put("/grade-step/{grade}/{step}")
async def put_grade_step(
    _user: ROLE_MASTER_ACCESS,
    grade: int,
    step: int,
    body: GradeStepUpdate,
    db: AsyncSession = Depends(get_db),
) -> dict:
    row = await master_service.update_grade_step(db, grade, step, body)
    return _ok(_grade_step_out(row))


@router.get("/allowance-rates")
async def get_allowance_rates(_user: AuthUser, db: AsyncSession = Depends(get_db)) -> dict:
    rows = await master_service.get_all_allowance_rates(db)
    return _ok([_allowance_out(r) for r in rows])


@router.post("/allowance-rates")
async def post_allowance_rate(
    _user: ROLE_MASTER_ACCESS,
    body: AllowanceRateCreate,
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    row = await master_service.create_allowance_rate(db, body)
    return JSONResponse(status_code=201, content=_ok(_allowance_out(row)))


@router.put("/allowance-rates/{name}")
async def put_allowance_rate(
    _user: ROLE_MASTER_ACCESS,
    name: str,
    body: AllowanceRateUpdate,
    db: AsyncSession = Depends(get_db),
) -> dict:
    row = await master_service.update_allowance_rate(db, name, body)
    return _ok(_allowance_out(row))


@router.get("/grade-derivation/derive")
async def get_grade_derivation_derive(
    _user: AuthUser,
    db: AsyncSession = Depends(get_db),
    education_level: str = Query(...),
    prior_experience_years: int = Query(...),
    years_of_service: int = Query(...),
) -> dict:
    result = await master_service.derive_grade(
        db,
        education_level,
        prior_experience_years,
        years_of_service,
    )
    return _ok(result)


@router.get("/grade-derivation")
async def get_grade_derivation(_user: AuthUser, db: AsyncSession = Depends(get_db)) -> dict:
    rows = await master_service.get_all_grade_derivations(db)
    return _ok([_grade_deriv_out(r) for r in rows])


@router.post("/grade-derivation")
async def post_grade_derivation(
    _user: ROLE_MASTER_ACCESS,
    body: GradeDerivationCreate,
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    row = await master_service.create_grade_derivation(db, body)
    return JSONResponse(status_code=201, content=_ok(_grade_deriv_out(row)))


@router.put("/grade-derivation/{edu}/{exp}")
async def put_grade_derivation(
    _user: ROLE_MASTER_ACCESS,
    edu: str,
    exp: int,
    body: GradeDerivationUpdate,
    db: AsyncSession = Depends(get_db),
) -> dict:
    row = await master_service.update_grade_derivation(db, edu, exp, body)
    return _ok(_grade_deriv_out(row))


@router.get("/org")
async def get_org(_user: AuthUser, db: AsyncSession = Depends(get_db)) -> dict:
    rows = await master_service.get_all_orgs(db)
    return _ok([_org_out(r) for r in rows])


@router.post("/org")
async def post_org(_user: ROLE_MASTER_ACCESS, body: OrgCreate, db: AsyncSession = Depends(get_db)) -> JSONResponse:
    row = await master_service.create_org(db, body)
    return JSONResponse(status_code=201, content=_ok(_org_out(row)))


@router.put("/org/{ministry_key}/{dept_key}")
async def put_org(
    _user: ROLE_MASTER_ACCESS,
    ministry_key: str,
    dept_key: str,
    body: OrgUpdate,
    db: AsyncSession = Depends(get_db),
) -> dict:
    row = await master_service.update_org(db, ministry_key, dept_key, body)
    return _ok(_org_out(row))


@router.get("/location")
async def get_location(_user: AuthUser, db: AsyncSession = Depends(get_db)) -> dict:
    rows = await master_service.get_all_locations(db)
    return _ok([_location_out(r) for r in rows])


@router.post("/location")
async def post_location(
    _user: ROLE_MASTER_ACCESS,
    body: LocationCreate,
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    row = await master_service.create_location(db, body)
    return JSONResponse(status_code=201, content=_ok(_location_out(row)))


@router.put("/location/{province_key}")
async def put_location(
    _user: ROLE_MASTER_ACCESS,
    province_key: str,
    body: LocationUpdate,
    db: AsyncSession = Depends(get_db),
) -> dict:
    row = await master_service.update_location(db, province_key, body)
    return _ok(_location_out(row))


@router.get("/bank")
async def get_bank(_user: AuthUser, db: AsyncSession = Depends(get_db)) -> dict:
    rows = await master_service.get_all_banks(db)
    return _ok([_bank_out(r) for r in rows])


@router.post("/bank")
async def post_bank(_user: ROLE_MASTER_ACCESS, body: BankCreate, db: AsyncSession = Depends(get_db)) -> JSONResponse:
    row = await master_service.create_bank(db, body)
    return JSONResponse(status_code=201, content=_ok(_bank_out(row)))


@router.put("/bank/{bank_name}/{bank_code}")
async def put_bank(
    _user: ROLE_MASTER_ACCESS,
    bank_name: str,
    bank_code: str,
    body: BankUpdate,
    db: AsyncSession = Depends(get_db),
) -> dict:
    row = await master_service.update_bank(db, bank_name, bank_code, body)
    return _ok(_bank_out(row))


@router.get("/pit-brackets")
async def get_pit_brackets(_user: AuthUser, db: AsyncSession = Depends(get_db)) -> dict:
    rows = await master_service.get_all_pit_brackets(db)
    return _ok([_pit_out(r) for r in rows])


@router.put("/pit-brackets/{bracket_no}")
async def put_pit_bracket(
    _user: ROLE_MASTER_ACCESS,
    bracket_no: int,
    body: PITBracketUpdate,
    db: AsyncSession = Depends(get_db),
) -> dict:
    row = await master_service.update_pit_bracket(db, bracket_no, body)
    return _ok(_pit_out(row))


@router.get("/employee-counts")
async def get_master_employee_counts(
    _user: ROLE_MASTER_ACCESS,
    db: AsyncSession = Depends(get_db),
) -> dict:
    async def _compute():
        return await master_service.compute_master_employee_counts(db)

    data = await cached_json(_MASTER_EMPLOYEE_COUNTS_KEY, _compute, ttl=300)
    return _ok(data)

File: frontend/src/api/master.ts
import { apiClient } from "@/api/client";
import type { ApiEnvelope } from "@/types/auth";

export interface GradeStep {
  grade: number;
  step: number;
  basic_salary: number;
}

export interface AllowanceRate {
  allowance_name: string;
  rate_type: string;
  rate_value: number;
  effective_date: string | null;
}

export interface AllowanceRateCreate {
  allowance_name: string;
  rate_type: string;
  rate_value: number;
  effective_date: string | null;
}

export interface GradeDerivation {
  education_level: string;
  min_exp_years: number;
  derived_grade: number;
  derived_step: number;
}

export interface GradeDerivationCreate {
  education_level: string;
  min_exp_years: number;
  exp_max_years?: number | null;
  derived_grade: number;
  derived_step: number;
}

export interface OrgRow {
  ministry_name: string;
  dept_key: string;
  dept_display_name: string;
  is_active: boolean;
}

export interface OrgCreate {
  ministry_name: string;
  dept_key: string;
  dept_display_name: string;
  is_active: boolean;
}

export interface Location {
  province: string;
  region: string;
  is_remote_area: boolean;
  is_hazardous_area: boolean;
  /** Present on rows from GET /master/location; omitted when creating. */
  is_active?: boolean;
}

export interface Bank {
  bank_name: string;
  bank_code: string;
  swift_code: string | null;
  is_active: boolean;
}

export interface BankUpdate {
  swift_code?: string | null;
  is_active?: boolean;
}

export interface PITBracket {
  bracket_no: number;
  lower_bound: number;
  upper_bound: number | null;
  rate_pct: number;
  deduction_amount: number;
}

export interface DeriveResult {
  grade: number;
  step: number;
  derived: boolean;
}

export async function fetchGradeSteps(): Promise<ApiEnvelope<GradeStep[]>> {
  const { data } = await apiClient.get<ApiEnvelope<GradeStep[]>>("/api/v1/master/grade-step");
  return data;
}

export async function updateGradeStep(
  grade: number,
  step: number,
  body: { basic_salary: number },
): Promise<ApiEnvelope<GradeStep>> {
  const { data } = await apiClient.put<ApiEnvelope<GradeStep>>(
    `/api/v1/master/grade-step/${grade}/${step}`,
    body,
  );
  return data;
}

export async function fetchAllowanceRates(): Promise<ApiEnvelope<AllowanceRate[]>> {
  const { data } = await apiClient.get<ApiEnvelope<AllowanceRate[]>>("/api/v1/master/allowance-rates");
  return data;
}

export async function createAllowanceRate(body: AllowanceRateCreate): Promise<ApiEnvelope<AllowanceRate>> {
  const { data } = await apiClient.post<ApiEnvelope<AllowanceRate>>("/api/v1/master/allowance-rates", body);
  return data;
}

export async function updateAllowanceRate(
  name: string,
  body: Partial<AllowanceRate>,
): Promise<ApiEnvelope<AllowanceRate>> {
  const { data } = await apiClient.put<ApiEnvelope<AllowanceRate>>(
    `/api/v1/master/allowance-rates/${encodeURIComponent(name)}`,
    body,
  );
  return data;
}

export async function fetchGradeDerivations(): Promise<ApiEnvelope<GradeDerivation[]>> {
  const { data } = await apiClient.get<ApiEnvelope<GradeDerivation[]>>("/api/v1/master/grade-derivation");
  return data;
}

export async function deriveGrade(params: {
  education_level: string;
  prior_experience_years: number;
  years_of_service: number;
}): Promise<ApiEnvelope<DeriveResult>> {
  const { data } = await apiClient.get<ApiEnvelope<DeriveResult>>("/api/v1/master/grade-derivation/derive", {
    params,
  });
  return data;
}

export async function createGradeDerivation(
  body: GradeDerivationCreate,
): Promise<ApiEnvelope<GradeDerivation>> {
  const { data } = await apiClient.post<ApiEnvelope<GradeDerivation>>("/api/v1/master/grade-derivation", body);
  return data;
}

export async function updateGradeDerivation(
  edu: string,
  exp: number,
  body: { derived_grade: number; derived_step: number },
): Promise<ApiEnvelope<GradeDerivation>> {
  const { data } = await apiClient.put<ApiEnvelope<GradeDerivation>>(
    `/api/v1/master/grade-derivation/${encodeURIComponent(edu)}/${exp}`,
    body,
  );
  return data;
}

export async function fetchOrgs(): Promise<ApiEnvelope<OrgRow[]>> {
  const { data } = await apiClient.get<ApiEnvelope<OrgRow[]>>("/api/v1/master/org");
  return data;
}

export async function createOrg(body: OrgCreate): Promise<ApiEnvelope<OrgRow>> {
  const { data } = await apiClient.post<ApiEnvelope<OrgRow>>("/api/v1/master/org", body);
  return data;
}

export async function updateOrg(
  ministry_name: string,
  dept_key: string,
  body: Partial<OrgRow>,
): Promise<ApiEnvelope<OrgRow>> {
  const { data } = await apiClient.put<ApiEnvelope<OrgRow>>(
    `/api/v1/master/org/${encodeURIComponent(ministry_name)}/${encodeURIComponent(dept_key)}`,
    body,
  );
  return data;
}

export async function fetchLocations(): Promise<ApiEnvelope<Location[]>> {
  const { data } = await apiClient.get<ApiEnvelope<Location[]>>("/api/v1/master/location");
  return data;
}

export async function createLocation(body: Location): Promise<ApiEnvelope<Location>> {
  const { data } = await apiClient.post<ApiEnvelope<Location>>("/api/v1/master/location", {
    province: body.province,
    region: body.region,
    is_remote_area: body.is_remote_area,
    is_hazardous_area: body.is_hazardous_area,
  });
  return data;
}

export async function updateLocation(
  province: string,
  body: Partial<Location>,
): Promise<ApiEnvelope<Location>> {
  const { data } = await apiClient.put<ApiEnvelope<Location>>(
    `/api/v1/master/location/${encodeURIComponent(province)}`,
    body,
  );
  return data;
}

export async function fetchBanks(): Promise<ApiEnvelope<Bank[]>> {
  const { data } = await apiClient.get<ApiEnvelope<Bank[]>>("/api/v1/master/bank");
  return data;
}

export async function createBank(body: Bank): Promise<ApiEnvelope<Bank>> {
  const { data } = await apiClient.post<ApiEnvelope<Bank>>("/api/v1/master/bank", {
    bank_name: body.bank_name,
    bank_code: body.bank_code,
    swift_code: body.swift_code ?? undefined,
    is_active: body.is_active,
  });
  return data;
}

export async function updateBank(
  bankName: string,
  bankCode: string,
  body: BankUpdate,
): Promise<ApiEnvelope<Bank>> {
  const { data } = await apiClient.put<ApiEnvelope<Bank>>(
    `/api/v1/master/bank/${encodeURIComponent(bankName)}/${encodeURIComponent(bankCode)}`,
    body,
  );
  return data;
}

export async function fetchPITBrackets(): Promise<ApiEnvelope<PITBracket[]>> {
  const { data } = await apiClient.get<ApiEnvelope<PITBracket[]>>("/api/v1/master/pit-brackets");
  return data;
}

export async function updatePITBracket(
  bracket_no: number,
  body: Partial<PITBracket>,
): Promise<ApiEnvelope<PITBracket>> {
  const { data } = await apiClient.put<ApiEnvelope<PITBracket>>(
    `/api/v1/master/pit-brackets/${bracket_no}`,
    body,
  );
  return data;
}
